src/utils/OrderedMsgChain.ts

```diff
@@ -16,133 +154,227 @@ const DEFAULT_PROPAGATION_TIMEOUT = 5000
-        })
-        /* eslint-enable arrow-body-style */
+        this.maxGapRequests = maxGapRequests
+
+        if (!this.isGapHandlingEnabled()) {
+            this.debug('Gap handling disabled for this %s.', this.constructor.name)
+        }
     }
 
+    /**
+     * Messages are stale if they are already enqueued or last ordered message is newer.
+     */
     isStaleMessage(streamMessage: StreamMessage) {
         const msgRef = streamMessage.getMessageRef()
         return (
-            this.lastReceivedMsgRef
-            && msgRef.compareTo(this.lastReceivedMsgRef) <= 0
+            // already enqueued
+            this.queue.has(streamMessage) || (
+                this.lastOrderedMsgRef
+                // or older/equal to last ordered msgRef
+                && msgRef.compareTo(this.lastOrderedMsgRef) <= 0
+            )
         )
     }
 
+    /**
+     * Add message to queue.
+     */
     add(unorderedStreamMessage: StreamMessage) {
         if (this.isStaleMessage(unorderedStreamMessage)) {
             const msgRef = unorderedStreamMessage.getMessageRef()
             // Prevent double-processing of messages for any reason
-            debug('Already received message: %o, lastReceivedMsgRef: %o. Ignoring message.', msgRef, this.lastReceivedMsgRef)
+            this.debug('Ignoring message: %o. Message was already enqueued or we already processed a newer message: %o.', msgRef, this.lastOrderedMsgRef)
             return
         }
 
-        if (this.isNextMessage(unorderedStreamMessage)) {
-            this.process(unorderedStreamMessage)
-        } else {
-            this.queue.push(unorderedStreamMessage)
+        // gap handling disabled
+        if (!this.isGapHandlingEnabled()) {
+            this.markMessage(unorderedStreamMessage)
         }
 
+        this.queue.push(unorderedStreamMessage)
         this.checkQueue()
     }
 
+    /**
+     * Mark a message to have it be treated as the next message & not trigger gap fill
+     */
     markMessageExplicitly(streamMessage: StreamMessage) {
+        // only add if not stale
+        if (this.markMessage(streamMessage)) {
+            this.add(streamMessage)
+        }
+    }
+
+    /**
+     * Adds message to set of marked messages.
+     * Does nothing and returns false if message is stale.
+     */
+    private markMessage(streamMessage: StreamMessage) {
         if (!streamMessage || this.isStaleMessage(streamMessage)) {
             // do nothing if already past/handled this message
-            return
+            return false
+        }
+
+        if (this.isGapHandlingEnabled()) {
+            this.debug('marking message', streamMessage.getMessageRef())
         }
 
         this.markedExplicitly.add(streamMessage)
-        this.add(streamMessage)
+        return true
     }
 
+    /**
+     * Cancel any outstanding gap fill request.
+     */
     clearGap() {
+        if (this.inProgress) {
+            this.debug('clearGap')
```

Comment:
clearGap is called after every message so it'll fill up the debug log without this.
I guess it could put the whole thing inside the same check but meh